File: android/app/src/main/java/com/sponteoai/chillscript/billing/AnnualPriceExperiment.kt
package com.sponteoai.chillscript.billing

import java.util.Locale

object AnnualPriceExperiment {
    const val ANNUAL_49_99 = "annual_49_99"
    const val ANNUAL_59_99 = "annual_59_99"
    const val OFFERING_49_99 = "post_login_two_page_trial_4999"
    const val OFFERING_59_99 = "post_login_two_page_trial_5999"
    const val BASE_PLAN_49_99 = "yearly49"
    const val BASE_PLAN_59_99 = "yearly"

    fun variantFor(userId: String): String {
        var hash = 0x811C9DC5L
        userId.lowercase(Locale.ROOT).toByteArray(Charsets.UTF_8).forEach { byte ->
            hash = hash xor (byte.toInt() and 0xFF).toLong()
            hash = (hash * 0x01000193L) and 0xFFFF_FFFFL
        }
        return if (hash % 100 < 50) ANNUAL_49_99 else ANNUAL_59_99
    }

    fun offeringIdentifierFor(userId: String): String = when (variantFor(userId)) {
        ANNUAL_49_99 -> OFFERING_49_99
        else -> OFFERING_59_99
    }

    fun annualBasePlanIdFor(userId: String): String = when (variantFor(userId)) {
        ANNUAL_49_99 -> BASE_PLAN_49_99
        else -> BASE_PLAN_59_99
    }
}

File: ios/chillnote/Services/AnnualPriceExperiment.swift
import Foundation

enum AnnualPriceExperiment: String {
    case annual4999 = "annual_49_99"
    case annual5999 = "annual_59_99"

    static let weeklyProductIdentifier = "com.chillnote.pro.weekly"

    var offeringIdentifier: String {
        switch self {
        case .annual4999: return "post_login_two_page_trial_4999"
        case .annual5999: return "post_login_two_page_trial_5999"
        }
    }

    var annualProductIdentifier: String {
        switch self {
        case .annual4999: return "com.chillnote.pro.yearly49"
        case .annual5999: return "com.chillnote.pro.yearly"
        }
    }

    var allowedProductIdentifiers: Set<String> {
        [Self.weeklyProductIdentifier, annualProductIdentifier]
    }

    static func variant(for userID: String) -> AnnualPriceExperiment {
        var hash: UInt32 = 2_166_136_261
        for byte in userID.lowercased().utf8 {
            hash ^= UInt32(byte)
            hash = hash &* 16_777_619
        }
        return hash % 100 < 50 ? .annual4999 : .annual5999
    }
}
